google-cloud-pubsub/lib/google/cloud/pubsub/publisher.rb

```diff
@@ -0,0 +1,205 @@
+          topic_name = service.topic_path topic_name
+          msg = create_pubsub_message data, attributes
+
+          synchronize do
+            batch = @queue[topic_name]
+            if batch.nil?
+              # Create a new batch.
+              batch = QueuedBatch.new msg, block
+            elsif !batch.try_add(msg, block, max_messages: max_messages,
+                                             max_bytes: max_bytes)
+              # We hit a threshold, publish existing batch.
+              publish_batch_async topic_name, batch
+              # Create a new batch.
+              batch = QueuedBatch.new msg, block
+            end
+            @queue[topic_name] = batch
+          end
+          nil
+        end
+
+        def start
+          @timer_task.execute
+        end
+
+        def stop
+          flush
+          @timer_task.shutdown
+        end
+
+        def started?
+          @timer_task.running?
+        end
+
+        def stopping?
+          @timer_task.shuttingdown?
+        end
+
+        def stopped?
+          @timer_task.shutdown?
+        end
+
+        def wait timeout = nil
+          @timer_task.wait_for_termination timeout
+        end
+
+        def flush
+          synchronize do
+            @queue.each do |topic_name, batch|
+              publish_batch_async topic_name, batch
+            end
+            @queue = {}
+          end
+        end
+
+        protected
+
+        def create_pubsub_message data, attributes
+          attributes ||= {}
+          if data.is_a?(::Hash) && attributes.empty?
+            attributes = data
+            data = nil
+          end
+
+          # Convert IO-ish objects to strings
+          if data.respond_to?(:read) && data.respond_to?(:rewind)
+            data.rewind
+            data = data.read
+          end
+          # Convert data to encoded byte array to match the protobuf defn
+          data = String(data).force_encoding("ASCII-8BIT")
+
+          # Convert attributes to strings to match the protobuf definition
+          attributes = Hash[attributes.map { |k, v| [String(k), String(v)] }]
+
+          Google::Pubsub::V1::PubsubMessage.new data: data,
+                                                attributes: attributes
+        end
+
+        def publish_batch topic_name, batch
+          grpc = service.publish_messages topic_name, batch.messages
+          batch.items.zip(Array(grpc.message_ids)) do |item, id|
+            item.msg.message_id = id
+            if item.callback
+              item.callback.call PublishResult.from_grpc(item.msg)
+            end
+          end
+        rescue => e
+          batch.items.each do |item|
+            if item.callback
+              item.callback.call PublishResult.from_error(item.msg, e)
+            end
+          end
+        end
+
+        def publish_batch_async topic_name, batch
+          Concurrent::Future.new(executor: @thread_pool) do
```

Comment:
You probably want an unbounded threadpool here -- each publish will consume an entire thread for the duration. gRPC's Ruby wrapper gives up the GVL before blocking, so we can make forward progress, as long as we have threads on which to perform work.